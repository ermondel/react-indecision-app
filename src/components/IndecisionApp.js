import React from 'react';
import Header from './Header';
import ActionBtn from './ActionBtn';
import OptionList from './OptionList';
import OptionModal from './OptionModal';
import {
  randomOption,
  addOption,
  removeOption,
  saveOptions,
  readOptions,
} from '../lib/optionList';

class IndecisionApp extends React.Component {
  state = {
    options: [],
    selectedOption: undefined,
    modalOpen: false,
    message: '',
  };

  subtitle = 'Put your life in the hands of a computer';

  showRandomOption = () => {
    this.__setState({
      selectedOption: randomOption(this.state.options),
      modalOpen: true,
    });
  };

  addOption = (option) => {
    if (!option) {
      this.__setState({ message: 'Enter valid value to add item' });
      return;
    }

    if (this.state.options.indexOf(option) >= 0) {
      this.__setState({ message: 'This option already exists' });
      return;
    }

    this.__setState({
      options: addOption(this.state.options, option),
    });
  };

  closeModal = () => {
    this.__setState({ modalOpen: false });
  };

  removeOption = (option) => {
    this.__setState({
      options: removeOption(this.state.options, option),
    });
  };

  removeAllOptions = () => {
    this.__setState({
      options: [],
    });
  };

  componentDidUpdate(prevProps, prevState) {
    saveOptions(this.state.options, prevState.options);
  }

  componentDidMount() {
    const options = readOptions();

    if (options) {
      this.__setState({ options });
    }
  }

  __setState(nextState) {
    if (!nextState.message && this.state.message) {
      nextState.message = '';
    }

    this.setState(nextState);
  }

  render() {
    return (
      <React.Fragment>
        <Header subtitle={this.subtitle} />

        <main className='main'>
          <div className='action-bar'>
            <ActionBtn
              disabled={this.state.options.length <= 0}
              onClick={this.showRandomOption}
            />
          </div>

          <OptionList
            options={this.state.options}
            addItem={this.addOption}
            removeItem={this.removeOption}
            removeAll={this.removeAllOptions}
            resultMessage={this.state.message}
          />
        </main>

        <OptionModal
          closeModal={this.closeModal}
          isOpen={this.state.modalOpen}
          selectedOption={this.state.selectedOption}
        />
      </React.Fragment>
    );
  }
}

export default IndecisionApp;
