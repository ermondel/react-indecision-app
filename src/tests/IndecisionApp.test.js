import React from 'react';
import renderer from 'react-test-renderer';
import IndecisionApp from '../components/IndecisionApp';
import Modal from 'react-modal';

jest.spyOn(Modal, "setAppElement").mockImplementation(() => {});

it('<IndecisionApp />', () => {
  const tree = renderer.create(<IndecisionApp />).toJSON();
  expect(tree).toMatchSnapshot();
});
