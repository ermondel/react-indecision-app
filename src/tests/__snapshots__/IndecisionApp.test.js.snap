// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndecisionApp /> 1`] = `
<main
  className="main"
>
  <div
    className="action-bar"
  >
    <button
      className="main-action-btn"
      disabled={true}
      onClick={[Function]}
    >
      What should I do?
    </button>
  </div>
  <div
    className="decisions"
  >
    <div
      className="decisions__header"
    >
      <h3
        className="decisions__title"
      >
        Your Options
      </h3>
    </div>
    <p
      className="decisions__list__message"
    >
      Please add an option to get started!
    </p>
    <div
      className="decisions__add"
    >
      
      <form
        className="decisions__add__form"
        onSubmit={[Function]}
      >
        <input
          className="decisions__add__input"
          name="decision"
          type="text"
        />
        <button
          className="decisions__add__submit-btn"
          name="submitbtn"
        >
          Add Option
        </button>
      </form>
    </div>
  </div>
</main>
`;
